File: leet/hashmap_set/difference_of_two_arrays_2215.py
from typing import List

class Solution:
    # ------------------------------------------------------------------------
    # 3ms
    # 未來嘗試先畫圖，有圖片後的思緒更清晰
    # 原本還有想說用two pointer
    def findDifference(self, nums1: List[int], nums2: List[int]) -> List[List[int]]:
        return [[list(set(nums1)-set(nums2))],[list(set(nums1)-set(nums2))]]

    # ------------------------------------------------------------------------
    # 要有排序才可以推動elif區塊，另一機制推動前進
    # list1 = sorted([3, 1, 4, 2])  # ➜ [1, 2, 3, 4]
    # list2 = sorted([4, 6, 5, 3])  # ➜ [3, 4, 5, 6]
    # while i < len(list1) and j < len(list2):
    # if list1[i] == list2[j]:
    #     result.append(list1[i])
    #     i += 1
    #     j += 1
    # elif list1[i] < list2[j]:
    #     i += 1
    # else:
    #     j += 1
    # ------------------------------------------------------------------------
    # 401ms
    # 大概是最慢的作法，O n square
    def findDifference(self, nums1: List[int], nums2: List[int]) -> List[List[int]]:
        result = [[],[]]
        for i in nums1:
            if i not in result[0] and i not in nums2:
                result[0].append(i)
        for j in nums2:
            if j not in result[1] and j not in nums1:
                result[1].append(j)
        return result

File: leet/hashmap_set/difference_of_two_arrays_2215.java
import java.util.*;

// ! 22ms
class Solution {
    public List<List<Integer>> findDifference(int[] nums1, int[] nums2) {
        Set<Integer> set1 = new HashSet<>(Arrays.stream(nums1).boxed().toList());
        Set<Integer> set2 = new HashSet<>(Arrays.stream(nums2).boxed().toList());

        Set<Integer> diff1 = new HashSet<>(set1);
        diff1.removeAll(set2);

        Set<Integer> diff2 = new HashSet<>(set2);
        diff2.removeAll(set1);

        return List.of(new ArrayList<>(diff1), new ArrayList<>(diff2));
        // 因為這邊出來的東西要是List<List<Integer>>，所以要轉換hash
    }
}

//
// -----------------------------------------------------------------------------
// ! removeAll()
// removeAll() 的回傳值是 boolean → 表示「有沒有成功移除任何元素」
// 但重點不是這個回傳值，而是它的副作用（副作用 = 直接改變了 diff1 本身）

// -----------------------------------------------------------------------------
// ! Set.copyOf
// Set<Integer> set1 = Set.copyOf(Arrays.stream(nums1).boxed().toList());
// Set<Integer> set2 = Set.copyOf(Arrays.stream(nums2).boxed().toList());
// 這用意是不讓複製過來的東西可改動，可以用再回傳的api

// -----------------------------------------------------------------------------
// ! 14ms
// class Solution {
// public List<List<Integer>> findDifference(int[] nums1, int[] nums2) {
// Set<Integer> set1 = new HashSet<>();
// Set<Integer> set2 = new HashSet<>();

// for (int num : nums1) set1.add(num);
// for (int num : nums2) set2.add(num);

// Set<Integer> diff1 = new HashSet<>(set1);
// diff1.removeAll(set2);

// Set<Integer> diff2 = new HashSet<>(set2);
// diff2.removeAll(set1);

// return List.of(new ArrayList<>(diff1), new ArrayList<>(diff2));
// }
// }
